dvc/stage/__init__.py

```diff
@@ -573,15 +573,21 @@ def run(
         if (self.cmd or self.is_import) and not self.frozen and not dry:
             self.remove_outs(ignore_remove=False, force=False)
 
-        if (not self.frozen and self.is_import) or self.is_partial_import:
+        if (
+            self.is_import and (not self.frozen or kwargs.get("pull"))
+        ) or self.is_partial_import:
             self._sync_import(dry, force, kwargs.get("jobs", None), no_download)
         elif not self.frozen and self.cmd:
             self._run_stage(dry, force, **kwargs)
-        else:
+        elif kwargs.get("pull"):
+            logger.info("Pulling data for %s", self)
+            for objs in self.get_used_objs().values():
+                self.repo.cloud.pull(objs)
```

Comment:
You probably need to pass odb here.
```suggestion
            for odb, objs in self.get_used_objs().items():
                self.repo.cloud.pull(objs, odb=odb)
```

Reply:
Talk to Ruslan and Peter please about this on how to pull from cloud versioned remotes. Most likely this involves index, but make sure to discuss. With this change, cloud versioning won't work.

Reply:
I got the reasoning for putting it inside here because of the rwlocks, but still missing the point of not using the high level pull API, as it would solve that

Reply:
@iterative/dvc could you reason what would be the downside of using `self.repo.pull` inside a stage vs trying to use some low-level API?

In addition to making this work for cloud versioning here, I also didn´t find a good way of pulling dependencies as in https://github.com/iterative/dvc/pull/9437#discussion_r1189920252 without using `stage.repo.pull`

Reply:
It's a large dependency that we are adding that is not meant for internal consumption. `Repo.pull()` is equivalent to `dvc pull`, so it's heavy and complex. And this one does that in a loop, for multiple stages.
`Repo.pull(stage.addressing)` reloads `dvc.yaml`/`dvc.lock` file, finds proper objects/index, and downloads them. 

Also, improving UI for `dvc repro`/`dvc exp run` may be difficult as it's now tied with `dvc pull`. What we really need here is a way to pull for one stage.
I don't have much issue with this as temporary solution, but I could see it getting permanent.

Reply:
Makes sense

> What we really need here is a way to pull for one stage.
I don't have much issue with this as temporary solution, but I could see it getting permanent.

Ok. I will open an issue to revisit this. Probably worth waiting for the new fetch